From cf76048eba5faa886745c03de32a2781af1a2bcf Mon Sep 17 00:00:00 2001
From: Patrick Barsanti <patrick.barsanti@amarulasolutions.com>
Date: Mon, 13 Jul 2026 17:14:27 +0200
Subject: [PATCH] avb: Add Rockchip patch to hardcode pubkey

Avoid using eFuses to verify the public key when checking Linux from
U-Boot.
This hack can be regarded as safe because the previous link of the chain
(U-Boot itself) will be verified using eFuses, therefore the U-Boot
source itself will not be able to be tampered with.

Note that I am not taking ownership of this hack/patch. It was taken
directly from the official public Rockchip documentation.

Upstream-Status: Inappropriate [oe-specific]
Signed-off-by: Patrick Barsanti <patrick.barsanti@amarulasolutions.com>
---
 common/android_bootloader.c        |  1 +
 include/android_avb/avb_root_pub.h | 52 ++++++++++++++++++++++++++++++
 lib/avb/libavb_user/avb_ops_user.c | 45 +++++++++++---------------
 3 files changed, 72 insertions(+), 26 deletions(-)
 create mode 100644 include/android_avb/avb_root_pub.h

diff --git a/common/android_bootloader.c b/common/android_bootloader.c
index 99059aef13..428afab267 100644
--- a/common/android_bootloader.c
+++ b/common/android_bootloader.c
@@ -827,6 +827,7 @@ static AvbSlotVerifyResult android_slot_verify(char *boot_partname,
 	if (ops->read_is_device_unlocked(ops, (bool *)&unlocked) != AVB_IO_RESULT_OK)
 		printf("Error determining whether device is unlocked.\n");
 
+	unlocked = 0; /* Lock State Fixed */
 	printf("read_is_device_unlocked() ops returned that device is %s\n",
 	       (unlocked & LOCK_MASK)? "UNLOCKED" : "LOCKED");
 
diff --git a/include/android_avb/avb_root_pub.h b/include/android_avb/avb_root_pub.h
new file mode 100644
index 0000000000..0cec062fb5
--- /dev/null
+++ b/include/android_avb/avb_root_pub.h
@@ -0,0 +1,52 @@
+#ifndef __AVB_ROOT_PUB_H_
+#define __AVB_ROOT_PUB_H_
+
+unsigned char avb_root_pub_bin[] = {
+  0x00, 0x00, 0x08, 0x00, 0xfc, 0xf2, 0xcc, 0xc7, 0x9c, 0xa7, 0xa1, 0xd0,
+  0x56, 0x8d, 0xed, 0xee, 0x4f, 0x90, 0xf5, 0x7c, 0xc0, 0xaa, 0x38, 0x82,
+  0xd9, 0xef, 0xab, 0xba, 0x96, 0x16, 0xdb, 0x78, 0xc0, 0x82, 0x63, 0x05,
+  0xb7, 0x24, 0x04, 0x12, 0x02, 0x63, 0x1a, 0xe3, 0xad, 0xef, 0x91, 0x4a,
+  0x7f, 0x71, 0x42, 0x37, 0xf8, 0x27, 0x1f, 0x0a, 0x60, 0x96, 0x73, 0x85,
+  0x12, 0xd1, 0xb9, 0x88, 0x3b, 0xa2, 0xf8, 0x07, 0x0f, 0xe9, 0x3d, 0xdd,
+  0xaf, 0xf8, 0x7c, 0xac, 0x29, 0x76, 0xc2, 0xc0, 0x8c, 0x03, 0x47, 0x83,
+  0xb0, 0x03, 0x95, 0x8b, 0xb1, 0x69, 0x35, 0x50, 0x6b, 0x08, 0xde, 0xb5,
+  0x17, 0x3f, 0x84, 0xcc, 0xe5, 0x83, 0x25, 0x78, 0xe4, 0x9a, 0xa8, 0x0c,
+  0x36, 0xe9, 0xe9, 0xbe, 0x82, 0x17, 0x6e, 0x0b, 0x7d, 0xdc, 0xdf, 0x3c,
+  0xc2, 0xa3, 0x5d, 0x06, 0xd5, 0xab, 0x6d, 0xfe, 0x12, 0x8d, 0xd0, 0x86,
+  0xe8, 0x87, 0x55, 0x89, 0x89, 0x6c, 0xd8, 0x16, 0x17, 0xda, 0xe0, 0x3a,
+  0xfa, 0x84, 0x73, 0x6b, 0x98, 0x7b, 0x8e, 0xdb, 0xbe, 0xac, 0x90, 0x11,
+  0xcb, 0x81, 0x7e, 0xaf, 0x59, 0xe5, 0x67, 0xcf, 0xcd, 0x6a, 0x85, 0xf9,
+  0x9a, 0x88, 0xb2, 0x97, 0xad, 0xad, 0x8b, 0x7e, 0x7f, 0x25, 0x84, 0xbc,
+  0x93, 0xa7, 0x52, 0xc2, 0xf4, 0x6d, 0x5a, 0x78, 0x2c, 0x78, 0x08, 0xcf,
+  0x7e, 0x40, 0xa7, 0x7a, 0x71, 0x95, 0x48, 0xed, 0xbc, 0x8c, 0x02, 0xc4,
+  0x32, 0xb5, 0x51, 0xa6, 0xca, 0xc8, 0x68, 0x45, 0xb7, 0xf1, 0xdb, 0xe5,
+  0x5b, 0x86, 0x26, 0x8c, 0x01, 0x48, 0xcd, 0xe1, 0x5a, 0x48, 0xa8, 0x23,
+  0xaa, 0x0e, 0x8b, 0x8f, 0xb7, 0x5c, 0xbe, 0xdd, 0xf2, 0x3e, 0x69, 0x6b,
+  0x19, 0xfb, 0xae, 0x64, 0x5d, 0x55, 0x2f, 0xfc, 0xb3, 0xf4, 0x5d, 0x93,
+  0xef, 0x10, 0xe6, 0x25, 0x49, 0x89, 0x81, 0x5d, 0x2d, 0x43, 0xcb, 0x09,
+  0x17, 0xd3, 0xdf, 0x1b, 0x49, 0xf4, 0xa6, 0x92, 0x2e, 0x94, 0x71, 0xc0,
+  0x69, 0x4d, 0xdf, 0xd4, 0x18, 0x97, 0xfe, 0x56, 0x99, 0x2d, 0xf1, 0xbb,
+  0x27, 0xa7, 0x6d, 0x4d, 0x0c, 0x3b, 0x2a, 0xc9, 0x26, 0xdd, 0xaf, 0xcb,
+  0xc5, 0x14, 0xea, 0xaa, 0x8b, 0xbf, 0xe4, 0x21, 0x8e, 0x79, 0xce, 0xf0,
+  0xf1, 0xc0, 0x3f, 0x33, 0x4b, 0x3e, 0xe4, 0x0c, 0x7f, 0x61, 0x14, 0x99,
+  0x56, 0x1e, 0x2f, 0xd8, 0x92, 0xf3, 0xc1, 0xa0, 0xff, 0xd1, 0xfd, 0x64,
+  0x62, 0x72, 0x34, 0x73, 0x1d, 0xb8, 0x28, 0x27, 0x82, 0x93, 0x22, 0xf9,
+  0x23, 0x42, 0xf0, 0x09, 0x5c, 0xb0, 0xa5, 0xb3, 0xe3, 0xe1, 0x49, 0xf5,
+  0xca, 0x79, 0xb3, 0x4e, 0xb6, 0x94, 0xd1, 0x04, 0x59, 0xde, 0x5c, 0x69,
+  0xdd, 0x53, 0x53, 0xfe, 0xf4, 0x6d, 0x7b, 0xbf, 0xc5, 0xab, 0x76, 0x7b,
+  0x6d, 0x23, 0x11, 0xd3, 0x7f, 0x91, 0x30, 0x5c, 0xb0, 0x5a, 0x9d, 0xbc,
+  0x67, 0x35, 0xc8, 0x5c, 0xa4, 0x3c, 0xbb, 0x5c, 0x54, 0xfb, 0xcf, 0xef,
+  0x02, 0x9e, 0x87, 0xb1, 0x92, 0xaa, 0x07, 0xc8, 0x78, 0xb0, 0xe1, 0xb7,
+  0x7c, 0x70, 0x00, 0x02, 0xbe, 0x18, 0x71, 0x76, 0x27, 0x61, 0x11, 0x01,
+  0x21, 0x60, 0xc7, 0x3d, 0x38, 0xe4, 0x45, 0xcb, 0xf9, 0x37, 0xe0, 0xd8,
+  0x66, 0x3e, 0x64, 0x67, 0x70, 0x12, 0x4d, 0x5a, 0x35, 0x66, 0x4b, 0x93,
+  0xaa, 0xd9, 0x8a, 0x1b, 0x19, 0xde, 0xd5, 0x2a, 0x8d, 0x48, 0xba, 0x87,
+  0xe9, 0xc6, 0x16, 0x5f, 0x21, 0xbf, 0x49, 0x94, 0x58, 0x29, 0xd4, 0x1e,
+  0x57, 0x0a, 0xf7, 0x68, 0xa7, 0xd8, 0x52, 0xd9, 0xca, 0x83, 0x84, 0x9f,
+  0x92, 0xb3, 0xd3, 0xbc, 0x3f, 0x8e, 0xe8, 0xf2, 0xc3, 0xb4, 0x6f, 0x00,
+  0x4f, 0x12, 0xb7, 0x5e, 0xe4, 0xa3, 0x33, 0x8d, 0x2d, 0x02, 0x57, 0xb0,
+  0xc2, 0x44, 0xbf, 0x30
+};
+unsigned int avb_root_pub_bin_len = 520;
+
+#endif /* __AVB_ROOT_PUB_H_ */
diff --git a/lib/avb/libavb_user/avb_ops_user.c b/lib/avb/libavb_user/avb_ops_user.c
index 4f57e7e617..2a05886123 100644
--- a/lib/avb/libavb_user/avb_ops_user.c
+++ b/lib/avb/libavb_user/avb_ops_user.c
@@ -42,6 +42,7 @@
 #include <android_avb/avb_vbmeta_image.h>
 #include <android_avb/avb_atx_validate.h>
 #include <boot_rkimg.h>
+#include <android_avb/avb_root_pub.h>
 
 static void byte_to_block(int64_t *offset,
 			  size_t *num_bytes,
@@ -192,21 +193,17 @@ validate_vbmeta_public_key(AvbOps *ops,
 			   size_t public_key_metadata_length,
 			   bool *out_is_trusted)
 {
-/* remain AVB_VBMETA_PUBLIC_KEY_VALIDATE to compatible legacy code */
-#if defined(CONFIG_AVB_VBMETA_PUBLIC_KEY_VALIDATE) || \
-    defined(AVB_VBMETA_PUBLIC_KEY_VALIDATE)
-	if (out_is_trusted) {
-		avb_atx_validate_vbmeta_public_key(ops,
-						   public_key_data,
-						   public_key_length,
-						   public_key_metadata,
-						   public_key_metadata_length,
-						   out_is_trusted);
+	if (!public_key_length || !out_is_trusted || !public_key_data) {
+		printf("%s: Invalid parameters\n", __func__);
+		return AVB_IO_RESULT_ERROR_IO;
 	}
-#else
-	if (out_is_trusted)
+
+	*out_is_trusted = false;
+	if (public_key_length != avb_root_pub_bin_len)
+		return AVB_IO_RESULT_ERROR_IO;
+	if (memcmp(public_key_data, avb_root_pub_bin, public_key_length) == 0)
 		*out_is_trusted = true;
-#endif
+
 	return AVB_IO_RESULT_OK;
 }
 
@@ -567,21 +564,17 @@ AvbIOResult validate_public_key_for_partition(AvbOps *ops,
 					      bool *out_is_trusted,
 					      uint32_t *out_rollback_index_location)
 {
-/* remain AVB_VBMETA_PUBLIC_KEY_VALIDATE to compatible legacy code */
-#if defined(CONFIG_AVB_VBMETA_PUBLIC_KEY_VALIDATE) || \
-    defined(AVB_VBMETA_PUBLIC_KEY_VALIDATE)
-	if (out_is_trusted) {
-		avb_atx_validate_vbmeta_public_key(ops,
-						   public_key_data,
-						   public_key_length,
-						   public_key_metadata,
-						   public_key_metadata_length,
-						   out_is_trusted);
+	if (!public_key_length || !out_is_trusted || !public_key_data) {
+		printf("%s: Invalid parameters\n", __func__);
+		return AVB_IO_RESULT_ERROR_IO;
 	}
-#else
-	if (out_is_trusted)
+
+	*out_is_trusted = false;
+	if (public_key_length != avb_root_pub_bin_len)
+		return AVB_IO_RESULT_ERROR_IO;
+	if (memcmp(public_key_data, avb_root_pub_bin, public_key_length) == 0)
 		*out_is_trusted = true;
-#endif
+
 	*out_rollback_index_location = 0;
 	return AVB_IO_RESULT_OK;
 }
